File: src/agent-src/templates/scripts/work_engine/hooks/context.py
"""``HookContext`` — payload carried into every hook callback.

One dataclass for both layers. Most fields are ``None`` for any given
event; the per-event subset is documented below and locked by the
roadmap's hook event surface table. Hooks must tolerate missing fields
gracefully — accessing a field that is ``None`` for the current event
is a hook bug, not an engine bug.

Per-event subset (mirrors the roadmap):

Dispatcher layer (``delivery`` is set; ``work`` is ``None``):
    - ``before_step``   → ``step_name``, ``delivery``
    - ``after_step``    → ``step_name``, ``delivery``, ``result``
    - ``on_halt``       → ``step_name``, ``delivery``, ``result``
    - ``on_error``      → ``step_name``, ``delivery``, ``exception``

CLI layer (``work`` is set; ``delivery`` may be set after load):
    - ``before_load``       → ``state_file``, ``args``
    - ``after_load``        → ``state_file``, ``work``, ``fmt``
    - ``before_dispatch``   → ``work``, ``delivery``, ``set_name``
    - ``after_dispatch``    → ``work``, ``delivery``, ``final``,
                              ``halting``
    - ``before_save``       → ``work``, ``delivery``, ``fmt``
    - ``after_save``        → ``work``, ``state_file``, ``fmt``
"""
from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
from typing import Any


@dataclass
class HookContext:
    """Per-event payload passed to every hook callback.

    Fields are intentionally optional — the runner does not validate
    which ones are populated for a given event. The contract is
    enforced by the call sites in ``dispatcher.py`` and ``cli.py``,
    not by the dataclass.

    ``extra`` exists as an escape hatch for hook-specific state that
    does not warrant a dedicated field. Use sparingly; if a piece of
    state is read by more than one hook, promote it to a real field.
    """

    # Dispatcher-layer refs.
    step_name: str | None = None
    delivery: Any = None  # DeliveryState — typed Any to avoid an import cycle.
    result: Any = None  # StepResult
    exception: BaseException | None = None

    # CLI-layer refs.
    work: Any = None  # WorkState — typed Any to avoid an import cycle.
    state_file: Path | None = None
    fmt: str | None = None
    set_name: str | None = None
    final: Any = None  # Outcome
    halting: str | None = None
    args: Any = None  # argparse.Namespace

    # Escape hatch for hook-specific state.
    extra: dict[str, Any] = field(default_factory=dict)


__all__ = ["HookContext"]

File: src/agent-src/templates/scripts/work_engine/hooks/context.ts
/**
 * `HookContext` — payload carried into every hook callback.
 *
 * TypeScript twin of `work_engine/hooks/context.py` (ADR-200 py2ts —
 * work_engine.hooks subpackage). One class for both layers. Most fields are
 * `null` for any given event; the per-event subset is documented below and
 * locked by the roadmap's hook event surface table. Hooks must tolerate
 * missing fields gracefully.
 *
 * Per-event subset (mirrors the roadmap):
 *
 * Dispatcher layer (`delivery` is set; `work` is `null`):
 *     - `before_step`   → `step_name`, `delivery`
 *     - `after_step`    → `step_name`, `delivery`, `result`
 *     - `on_halt`       → `step_name`, `delivery`, `result`
 *     - `on_error`      → `step_name`, `delivery`, `exception`
 *
 * CLI layer (`work` is set; `delivery` may be set after load):
 *     - `before_load`       → `state_file`, `args`
 *     - `after_load`        → `state_file`, `work`, `fmt`
 *     - `before_dispatch`   → `work`, `delivery`, `set_name`
 *     - `after_dispatch`    → `work`, `delivery`, `final`, `halting`
 *     - `before_save`       → `work`, `delivery`, `fmt`
 *     - `after_save`        → `work`, `state_file`, `fmt`
 */

/** Arbitrary value, mirroring the Python `Any` fields. */
export type Any = unknown;

/** Per-event payload passed to every hook callback (matches the dataclass). */
export interface HookContextInit {
    // Dispatcher-layer refs.
    step_name?: string | null;
    delivery?: Any; // DeliveryState
    result?: Any; // StepResult
    exception?: Any; // BaseException

    // CLI-layer refs.
    work?: Any; // WorkState
    state_file?: string | null;
    fmt?: string | null;
    set_name?: string | null;
    final?: Any; // Outcome
    halting?: string | null;
    args?: Any; // argparse.Namespace

    // Escape hatch for hook-specific state.
    extra?: Record<string, Any>;
}

/**
 * Per-event payload passed to every hook callback.
 *
 * Fields are intentionally optional — the runner does not validate which
 * ones are populated for a given event. The contract is enforced by the
 * call sites, not by the class.
 *
 * `extra` exists as an escape hatch for hook-specific state that does not
 * warrant a dedicated field.
 */
export class HookContext {
    // Dispatcher-layer refs.
    step_name: string | null;
    delivery: Any;
    result: Any;
    exception: Any;

    // CLI-layer refs.
    work: Any;
    state_file: string | null;
    fmt: string | null;
    set_name: string | null;
    final: Any;
    halting: string | null;
    args: Any;

    // Escape hatch for hook-specific state.
    extra: Record<string, Any>;

    constructor(init: HookContextInit = {}) {
        this.step_name = init.step_name ?? null;
        this.delivery = init.delivery ?? null;
        this.result = init.result ?? null;
        this.exception = init.exception ?? null;
        this.work = init.work ?? null;
        this.state_file = init.state_file ?? null;
        this.fmt = init.fmt ?? null;
        this.set_name = init.set_name ?? null;
        this.final = init.final ?? null;
        this.halting = init.halting ?? null;
        this.args = init.args ?? null;
        this.extra = init.extra ?? {};
    }
}
